### PR TITLE
Support Tasks that perform a single iteration

### PR DESCRIPTION
**Pull request by adrianna-chang-shopify**
Closes: https://github.com/Shopify/maintenance_tasks/issues/568

Sometimes, a Task may simply need to perform an action without iterating over multiple elements in a collection. For example, a Task might need to kick off another job, interact with a third-party API, or perform a single operation. This adds support for "collection-less" tasks, which need only to define a `#process` method.

I'm proposing an API similar to what we've done for `csv_collection` tasks, where we use a `no_collection` macro in the task. Under the hood, we can use a `NoCollectionBuilder` to handle defining `#collection` and `#count` (with collection simply being `:no_collection` for use when building the enumerator), and in the job we can use Job Iteration's `#build_once_enumerator` to handle this.

I've added a generator option for this as a distinct boolean option. We could change the task generator to take a `type` option (ie. `--type=csv`, `--type=no-collection`) but this is technically an API change, and I'm not sure it's necessary to generalize this yet (maybe it will make more sense if we have _three_ options at some point in the future 😉 ). Instead, I've opted to raise if a user supplies multiple option flags.

## Feedback

Thoughts on the generator flag / macro name? Is `process_once` more suitable?

### CODE REVIEW

File: lib/generators/maintenance_tasks/task_generator.rb [hunk 1]
```diff
@@ -12,17 +12,26 @@ class TaskGenerator < Rails::Generators::NamedBase
     class_option :csv, type: :boolean, default: false,
       desc: "Generate a CSV Task."
 
+    class_option :no_collection, type: :boolean, default: false,
```
Comment (gmcgibbon) [inline lib/generators/maintenance_tasks/task_generator.rb:15]: To me, no_collection doesn't really imply a single run. Maybe `once`, `single_run` or `one_off` would make more sense?

File: app/jobs/concerns/maintenance_tasks/task_job_concern.rb [hunk 2]
```diff
@@ -36,6 +36,8 @@ def build_enumerator(_run, cursor:)
       @enumerator = nil
 
       collection_enum = case collection
+      when :no_collection
+        enumerator_builder.build_once_enumerator(cursor: nil)
```
Comment (gmcgibbon) [inline app/jobs/concerns/maintenance_tasks/task_job_concern.rb:40]: https://github.com/Shopify/job-iteration/blob/53767c94295af5655971c3b33397da692227519f/lib/job-iteration/enumerator_builder.rb#L40

I believe that means that `1` will be passed to the enumerator, if anyone else was curious.

**Review by gmcgibbon — APPROVED** [review]

File: app/jobs/concerns/maintenance_tasks/task_job_concern.rb (hunk 2 above)
Comment (adrianna-chang-shopify) [reply app/jobs/concerns/maintenance_tasks/task_job_concern.rb:40]: > I believe that means that 1 will be passed to the enumerator

Yes! I was also a bit intrigued when I saw this. Indeed, that means that `1` is passed to the user's `process` method, but I don't think it's a big deal either 😄

File: lib/generators/maintenance_tasks/task_generator.rb (hunk 1 above)
Comment (gmcgibbon) [reply lib/generators/maintenance_tasks/task_generator.rb:15]: I mean keeping
```rb
def collection(_task)
  :no_collection
end
```
and `when :no_collection` but renaming the class option, builder class, and class macro method to `single_run`. `process_once` or `one_off` seem fine to me as well, no strong opinion on which is best.

**Comment by adrianna-chang-shopify** [comment]
> I wish we could solve the issue that process takes a parameter though.

Yeah agreed @etiennebarrie 😄 I thought about doing something like this:

```diff
diff --git a/app/jobs/concerns/maintenance_tasks/task_job_concern.rb b/app/jobs/concerns/maintenance_tasks/task_job_concern.rb
index 59c09d0..f885984 100644
--- a/app/jobs/concerns/maintenance_tasks/task_job_concern.rb
+++ b/app/jobs/concerns/maintenance_tasks/task_job_concern.rb
@@ -91,7 +91,11 @@ module MaintenanceTasks
     end
 
     def task_iteration(input)
-      @task.process(input)
+      if @task.no_collection_task
+        @task.process
+      else
+        @task.process(input)
+      end
     rescue => error
       @errored_element = input
       raise error
diff --git a/app/models/maintenance_tasks/task.rb b/app/models/maintenance_tasks/task.rb
index ce33e83..384ac31 100644
--- a/app/models/maintenance_tasks/task.rb
+++ b/app/models/maintenance_tasks/task.rb
@@ -21,6 +21,8 @@ module MaintenanceTasks
     class_attribute :collection_builder_strategy,
       default: NullCollectionBuilder.new
 
+    class_attribute :no_collection_task, default: false
+
     define_callbacks :start, :complete, :error, :cancel, :pause, :interrupt
 
     class << self
@@ -68,6 +70,7 @@ module MaintenanceTasks
       def no_collection
         self.collection_builder_strategy =
           MaintenanceTasks::NoCollectionBuilder.new
+        self.no_collection_task = true
       end
 
       # Returns whether the Task handles CSV.
diff --git a/test/dummy/app/tasks/maintenance/no_collection_task.rb b/test/dummy/app/tasks/maintenance/no_collection_task.rb
index 67f2bb3..ef4dcbb 100644
--- a/test/dummy/app/tasks/maintenance/no_collection_task.rb
+++ b/test/dummy/app/tasks/maintenance/no_collection_task.rb
@@ -4,7 +4,7 @@ module Maintenance
   class NoCollectionTask < MaintenanceTasks::Task
     no_collection
 
-    def process(_)
+    def process
       Rails.logger.debug("#process method was called")
     end
   end
```

But wasn't sure if the added complexity was worth it given the generator gives you a template to follow. It's really not that much code though 😄  And I do wonder about the generated test suit:
```
class <%= class_name %>TaskTest < ActiveSupport::TestCase
  # test "#process performs a task iteration" do
  #   <%= tasks_module %>::<%= class_name %>Task.process(element)
  # end
end
```
I suppose this is potentially confusing? WDYT about going with the suggestion above ^ and providing a new template for collection-less tasks?

**Review by nvasilevski — APPROVED** [review]
I like the feature 👍 
Personally I feel it's reasonable that we are going with a macro approach to be consistent with `csv_collection`. 
Naming is hard so I don't have a strong preference, `no_collection` seems fine to me :) 
I agree that ideally we would want to have an argument-less `process` method, but I think it's not a blocker for the feature in general? I was thinking that perhaps we could have a separate PR for tweaking the `process` method to evaluate reasonability and possible approaches. 
It's just this particular PR looks nice and clean to me and it makes a decent UX improvement in terms of moving from:
```ruby
def collection
 [1]
end

def process(_)
end
```
to

```ruby
no_collection

def process(_)
end
```

**Pull request merged**